python/tank/platform/engine.py

```diff
@@ -421,6 +433,29 @@ def post_app_init(self):
         Implemented by deriving classes.
         """
         pass
+
+    def pre_context_change(self, old_context, new_context):
```

Comment:
Same thought here: Maybe `Bundle` should implement `context_change_allowed, change_context, pre/post_context_change` ?